### PR TITLE
WIP: experiment with the midpoint method. DO NOT MERGE

### PR DESCRIPTION
**Pull request by martynpclark**
Initial testing of the midpoint method.

### CODE REVIEW

File: build/source/dshare/globalData.f90 [hunk 1]
```diff
@@ -55,6 +55,10 @@ MODULE globalData
  ! * part 1: parameters that are fixed across multiple instantiations
  ! ----------------------------------------------------------------------------------------------------------------
 
+ ! define version number
+ character(len=256),parameter,public         :: summaVersion='v2.34.56'
+`
+
```
Comment (martynpclark) [inline build/source/dshare/globalData.f90]: Why is length=256. Shouldn't this be done in globalData?

File: build/source/dshare/globalData.f90 [hunk 2]
```diff
@@ -55,6 +55,10 @@ MODULE globalData
  ! * part 1: parameters that are fixed across multiple instantiations
  ! ----------------------------------------------------------------------------------------------------------------
 
+ ! define version number
+ character(len=256),parameter,public         :: summaVersion='v2.34.56'
+`
```
Comment (martynpclark) [inline build/source/dshare/globalData.f90]: This does not have anything to do with implementing the midpoint method. Please separate the tasks into independent PRs.

Comment (martynpclark) [inline build/source/dshare/globalData.f90]: Where did v2.34.56 come from? Is this an actual version number? Please check.

Comment (martynpclark) [inline build/source/dshare/globalData.f90]: Should the summa version also be included in control files? This implementation is not correct.

File: build/source/engine/systemSolv.f90 [hunk 3]
```diff
@@ -217,12 +217,16 @@ real(dp)                        :: fScale(nState)                ! characteristi
  real(qp)                        :: rVec(nState)    ! NOTE: qp    ! residual vector
  real(dp)                        :: rAdd(nState)                  ! additional terms in the residual vector
  real(dp)                        :: fOld,fNew                     ! function values (-); NOTE: dimensionless because scaled
- real(dp)                        :: xMin,xMax                     ! state minimum and maximum (mixed units) 
+ real(dp)                        :: xMin,xMax                     ! state minimum and maximum (mixed units)
```
Comment (martynpclark) [inline build/source/engine/systemSolv.f90]: This is just a change in white space. Please ensure that all code has been run through a white space trimmer before review.

File: build/source/engine/systemSolv.f90 [hunk 4]
```diff
@@ -217,12 +217,16 @@ real(dp)                        :: fScale(nState)                ! characteristi
...
  logical(lgt)                    :: converged                     ! convergence flag
  logical(lgt)                    :: feasible                      ! feasibility flag
  real(dp)                        :: resSinkNew(nState)            ! additional terms in the residual vector
  real(dp)                        :: fluxVecNew(nState)            ! new flux vector
  real(qp)                        :: resVecNew(nState)  ! NOTE: qp ! new residual vector
+ integer(i4b), parameter         :: firstOrder=8991               ! named variable for the first order solution
+ integer(i4b), parameter         :: secondOrder=8992              ! named variable for the second order solution
+ integer(i4b)                    :: ixSolution=firstOrder         ! decision for the solution method
+ real(dp)                        :: dtAdj                         ! adjusted time step (s)
```
Comment (martynpclark) [inline build/source/engine/systemSolv.f90]: should this be included in the decisions file? Hard coded at the moment.

File: build/source/engine/systemSolv.f90 [hunk 5]
```diff
@@ -348,6 +352,13 @@ fScale,                           & ! intent(out):   function scaling vector (mi
  ! * compute the initial function evaluation...
  ! --------------------------------------------
 
+ ! define time step
+ select case(ixSolution)
+  case(firstOrder);  dtAdj = dt       ! first order solution
+  case(secondOrder); dtAdj = dt/2._dp ! second order solution
+  case default; err=20; message=trim(message)//'unable to identify solution method'; return
+ end select ! (define time step)
+
```
Comment (martynpclark) [inline build/source/engine/systemSolv.f90]: Good.

File: build/source/engine/systemSolv.f90 [hunk 6]
```diff
@@ -365,7 +376,8 @@ fScale,                           & ! intent(out):   function scaling vector (mi
  ! NOTE 2: The Jacobian matrix together with the residual vector is used to calculate the first iteration increment
  call eval8summa(&
                  ! input: model control
-                 dt,                      & ! intent(in):    length of the time step (seconds)
+                 !dt,                      & ! intent(in):    length of the time step (seconds)
```
Comment (martynpclark) [inline build/source/engine/systemSolv.f90]: should not have commented values in the PR. Please remove the commented line..

File: build/source/engine/systemSolv.f90 [hunk 7]
```diff
@@ -535,6 +548,9 @@ fNew,                          & ! intent(out):   new function evaluation
  ! set untapped melt energy to zero
  untappedMelt(:) = 0._dp
 
+ ! estimate states at the end of the time step
+ if(ixSolution==secondOrder) stateVecTrial = stateVecTrial(:)*2._dp - stateVecInit(:)
+
```
Comment (martynpclark) [inline build/source/engine/systemSolv.f90]: good

**Comment by bartnijssen** [comment]
Closing this PR for now. Can be reopened as need be after we release V3.0.0 (and that version should be merged into this branch first)
